classes/CartRule.php

```diff
@@ -1005,6 +1005,25 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
         if ($check_carrier) {
             $otherCartRules = $cart->getCartRules(CartRule::FILTER_ACTION_ALL, false);
         }
+
+        // Check if the gift product is available for order
+        if ($this->gift_product) {
+            $shopId = (int) Context::getContext()->shop->id;
+            $availableQuantity = StockAvailable::getQuantityAvailableByProduct(
```

Comment:
Can you check if a pack product can be added as a gift? If yes, Product::getQuantity must be used, because it properly accounts for packs.

Otherwise, good job!